### PR TITLE
feat: implement eip4788

### PR DESCRIPTION
**Pull request by terencechain**
This PR implements [EIP4788](https://github.com/ethereum/consensus-specs/pull/3421) on Prysm. Majority of the changes are on the [execution API](https://github.com/ethereum/execution-apis/pull/420)

### CODE REVIEW

File: consensus-types/payload-attribute/getters.go [hunk 1]
```diff
@@ -74,3 +74,23 @@ func (a *data) PbV2() (*enginev1.PayloadAttributesV2, error) {
...
+// PbV3 returns the payload attribute in version 3.
+func (a *data) PbV3() (*enginev1.PayloadAttributesV3, error) {
+	if a == nil {
+		return nil, errNilPayloadAttribute
+	}
+	if a.version != version.Deneb {
+		return nil, consensus_types.ErrNotSupported("PayloadAttributePbV3", a.version)
+	}
+	if a.timeStamp == 0 && len(a.prevRandao) == 0 {
+		return nil, nil
```
Comment (rauljordan) [inline consensus-types/payload-attribute/getters.go:87]: Given we are validating fields here, should we check that parent beacon block root and other fields are non-nil here?

Comment (rauljordan) [reply consensus-types/payload-attribute/getters.go:87]: and non-empty

Comment (terencechain) [reply consensus-types/payload-attribute/getters.go:87]: I dont think this check is necessary so ill remove it

**Review by saolyn — COMMENTED** [review]
You seem to have a couple messed up imports, probably from another branch that was merged in

File: consensus-types/payload-attribute/getters.go [hunk 2]
```diff
@@ -74,3 +74,20 @@ func (a *data) PbV2() (*enginev1.PayloadAttributesV2, error) {
...
 	}, nil
 }
+
+// PbV3 returns the payload attribute in version 3.
+func (a *data) PbV3() (*enginev1.PayloadAttributesV3, error) {
+	if a == nil {
+		return nil, errNilPayloadAttribute
+	}
+	if a.version != version.Deneb {
+		return nil, consensus_types.ErrNotSupported("PayloadAttributePbV3", a.version)
```
Comment (rkapka) [inline consensus-types/payload-attribute/getters.go:84]:
```suggestion
		return nil, consensus_types.ErrNotSupported("PbV3", a.version)
```
Previous versions are also mismatched.

File: consensus-types/payload-attribute/getters_test.go [hunk 3]
```diff
@@ -135,6 +146,42 @@ func TestPayloadAttributeGetters(t *testing.T) {
...
+				require.Equal(t, (*enginev1.PayloadAttributesV3)(nil), got)
+			},
+		},
+		{
+			name: "Get PbDeneb (bad version)",
+			tc: func(t *testing.T) {
+				a, err := New(&enginev1.PayloadAttributesV2{})
+				require.NoError(t, err)
+				_, err = a.PbV3()
+				require.ErrorContains(t, "PayloadAttributePbV3 is not supported for capella: unsupported getter", err)
```
Comment (rkapka) [inline consensus-types/payload-attribute/getters_test.go:182]: This is not a suggestion to change the code, but my personal opinion is that tests are less fragile when we check for a small meaningful part of the error message that is unlikely to change. In this example I would only check for `not supported`. Having the whole error message in the assertion means that you will need to update the test if the name of the function changes, or the `unsupported getter` part.

Comment (terencechain) [reply consensus-types/payload-attribute/getters_test.go:182]: I agree, but it's also good to check against `PbV3` vs `V2` or `V1`. In this case, I am inclined to leave it as it is

Comment (terencechain) [reply consensus-types/payload-attribute/getters_test.go:182]: I agree, but it's also good to check against `PbV3` vs `V2` or `V1`. In this case, I am inclined to leave it as it is

**Comment by terencechain** [comment]
ready for another look @rkapka @rauljordan !

**Comment by terencechain** [comment]
e2e failed, we'll need https://github.com/ethereum/go-ethereum/pull/27289 to merge

**Review by james-prysm — APPROVED** [review]
I thought this was good i didn't find anything that stood out.

File: proto/engine/v1/execution_engine.proto [hunk 4]
```diff
@@ -173,6 +173,14 @@ message PayloadAttributesV2 {
...
 }
 
+message PayloadAttributesV3 {
+	uint64 timestamp              = 1;
+	bytes prev_randao             = 2 [(ethereum.eth.ext.ssz_size) = "32"];
+	bytes suggested_fee_recipient = 3 [(ethereum.eth.ext.ssz_size) = "20"];
+	repeated Withdrawal withdrawals = 4 [(ethereum.eth.ext.ssz_max) = "withdrawal.size"];
+	bytes parent_beacon_block_root             = 5 [(ethereum.eth.ext.ssz_size) = "32"];
+}
+
```
Comment (prestonvanloon) [inline proto/engine/v1/execution_engine.proto:183]: please fix formatting.

I believe clang-format works https://clang.llvm.org/docs/ClangFormat.html

File: beacon-chain/blockchain/execution_engine_test.go [hunk 5]
```diff
@@ -1100,9 +1106,9 @@ func TestKZGCommitmentToVersionedHashes(t *testing.T) {
 	require.NoError(t, err)
 	vhs, err := kzgCommitmentsToVersionedHashes(b.Block().Body())
 	require.NoError(t, err)
-	vh0 := [32]byte{1, 207, 35, 21, 201, 118, 88, 167, 237, 84, 173, 161, 129, 118, 94, 35, 179, 250, 219, 81, 80, 250, 179, 149, 9, 246, 49, 192, 185, 175, 69, 102}
-	vh1 := [32]byte{1, 226, 124, 226, 142, 82, 126, 176, 113, 150, 179, 26, 240, 245, 250, 24, 130, 172, 231, 1, 166, 130, 2, 42, 183, 121, 248, 22, 172, 57, 212, 126}
+	vh0 := "0x01cf2315c97658a7ed54ada181765e23b3fadb5150fab39509f631c0b9af4566"
+	vh1 := "0x01e27ce28e527eb07196b31af0f5fa1882ace701a682022ab779f816ac39d47e"
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/execution_engine_test.go:1110]: yeah this is a nice improvement

**Pull request merged**